### ISSUE TITLE
OIDC Custom Root CA - unable to get local issuer certificate

### ISSUE THREAD

**Issue by thenetworkdoctor** [pschiffe/docker-pdns#143]
Hi, 

I'm running the docker-compose-pgsql DNS stack (I only added an traefik reverse proxy for handling the SSL termination for pdns-admin. (no issues with setting this up, or managing this)

I'm also running my own PKI, with a custom root CA (distributed to all clients), this root is also used for all SSL traffic in the environment.

When setting up OIDC between pdns-admin and my authentik (running on another host: "authentik.lab.internal"), I ran into an error when clicking the "Sign in with oidc" button, the web page shows a http error 500.

The docker compose logs show an SSL: CERTIFICATE_VERIFY_FAILED
(snipped the most relevant logs (i think)

I'm suspecting the root CA not being available in the container (pdns-admin) but i don't know (can't find)
 if there is an environment variable or volume i can map for this.

Any idea how to fix this?

maybe unrelated; but i'm just spitting here: This environment i skipped the LDAP setup, but i'm guessing that is going to experience the same issues when using ldaps.

```
pdns-admin-pgsql-1     | [2024-09-19 15:02:25,403] [app.py:1449] ERROR - Exception on /oidc/login [GET]
pdns-admin-pgsql-1     | Traceback (most recent call last):
pdns-admin-pgsql-1     |   File "/usr/local/lib/python3.9/site-packages/urllib3/connectionpool.py", line 715, in urlopen
pdns-admin-pgsql-1     |     httplib_response = self._make_request(
pdns-admin-pgsql-1     |   File "/usr/local/lib/python3.9/site-packages/urllib3/connectionpool.py", line 404, in _make_request
pdns-admin-pgsql-1     |     self._validate_conn(conn)
pdns-admin-pgsql-1     |   File "/usr/local/lib/python3.9/site-packages/urllib3/connectionpool.py", line 1058, in _validate_conn
pdns-admin-pgsql-1     |     conn.connect()
pdns-admin-pgsql-1     |   File "/usr/local/lib/python3.9/site-packages/urllib3/connection.py", line 419, in connect
pdns-admin-pgsql-1     |     self.sock = ssl_wrap_socket(
pdns-admin-pgsql-1     |   File "/usr/local/lib/python3.9/site-packages/urllib3/util/ssl_.py", line 449, in ssl_wrap_socket
pdns-admin-pgsql-1     |     ssl_sock = _ssl_wrap_socket_impl(
pdns-admin-pgsql-1     |   File "/usr/local/lib/python3.9/site-packages/urllib3/util/ssl_.py", line 493, in _ssl_wrap_socket_impl
pdns-admin-pgsql-1     |     return ssl_context.wrap_socket(sock, server_hostname=server_hostname)
pdns-admin-pgsql-1     |   File "/usr/lib64/python3.9/ssl.py", line 501, in wrap_socket
pdns-admin-pgsql-1     |     return self.sslsocket_class._create(
pdns-admin-pgsql-1     |   File "/usr/lib64/python3.9/ssl.py", line 1074, in _create
pdns-admin-pgsql-1     |     self.do_handshake()
pdns-admin-pgsql-1     |   File "/usr/lib64/python3.9/ssl.py", line 1343, in do_handshake
pdns-admin-pgsql-1     |     self._sslobj.do_handshake()
pdns-admin-pgsql-1     | ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1129)
pdns-admin-pgsql-1     | 
pdns-admin-pgsql-1     | During handling of the above exception, another exception occurred:
pdns-admin-pgsql-1     | 
pdns-admin-pgsql-1     | Traceback (most recent call last):
pdns-admin-pgsql-1     |   File "/usr/local/lib/python3.9/site-packages/requests/adapters.py", line 489, in send
pdns-admin-pgsql-1     |     resp = conn.urlopen(
pdns-admin-pgsql-1     |   File "/usr/local/lib/python3.9/site-packages/urllib3/connectionpool.py", line 799, in urlopen
pdns-admin-pgsql-1     |     retries = retries.increment(
pdns-admin-pgsql-1     |   File "/usr/local/lib/python3.9/site-packages/urllib3/util/retry.py", line 592, in increment
pdns-admin-pgsql-1     |     raise MaxRetryError(_pool, url, error or ResponseError(cause))
pdns-admin-pgsql-1     | urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='authentik.lab.internal', port=443): Max retries exceeded with url: /application/o/powerdns-admin/.well-known/openid-configuration (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1129)')))
pdns-admin-pgsql-1     | 
pdns-admin-pgsql-1     | During handling of the above exception, another exception occurred:
pdns-admin-pgsql-1     | 
pdns-admin-pgsql-1     | Traceback (most recent call last):
pdns-admin-pgsql-1     |   File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 2073, in wsgi_app
pdns-admin-pgsql-1     |     response = self.full_dispatch_request()
pdns-admin-pgsql-1     |   File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1519, in full_dispatch_request
pdns-admin-pgsql-1     |     rv = self.handle_user_exception(e)
pdns-admin-pgsql-1     |   File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1517, in full_dispatch_request
pdns-admin-pgsql-1     |     rv = self.dispatch_request()
pdns-admin-pgsql-1     |   File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1503, in dispatch_request
pdns-admin-pgsql-1     |     return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
pdns-admin-pgsql-1     |   File "/opt/powerdns-admin/powerdnsadmin/routes/index.py", line 155, in oidc_login
pdns-admin-pgsql-1     |     return oidc.authorize_redirect(redirect_uri)
pdns-admin-pgsql-1     |   File "/usr/local/lib/python3.9/site-packages/authlib/integrations/flask_client/apps.py", line 43, in authorize_redirect
pdns-admin-pgsql-1     |     rv = self.create_authorization_url(redirect_uri, **kwargs)
pdns-admin-pgsql-1     |   File "/usr/local/lib/python3.9/site-packages/authlib/integrations/base_client/sync_app.py", line 309, in create_authorization_url
pdns-admin-pgsql-1     |     metadata = self.load_server_metadata()
pdns-admin-pgsql-1     |   File "/usr/local/lib/python3.9/site-packages/authlib/integrations/base_client/sync_app.py", line 294, in load_server_metadata
pdns-admin-pgsql-1     |     resp = session.request('GET', self._server_metadata_url, withhold_token=True)
pdns-admin-pgsql-1     |   File "/usr/local/lib/python3.9/site-packages/authlib/integrations/requests_client/oauth2_session.py", line 109, in request
pdns-admin-pgsql-1     |     return super(OAuth2Session, self).request(
pdns-admin-pgsql-1     |   File "/usr/local/lib/python3.9/site-packages/requests/sessions.py", line 587, in request
pdns-admin-pgsql-1     |     resp = self.send(prep, **send_kwargs)
pdns-admin-pgsql-1     |   File "/usr/local/lib/python3.9/site-packages/requests/sessions.py", line 701, in send
pdns-admin-pgsql-1     |     r = adapter.send(request, **kwargs)
pdns-admin-pgsql-1     |   File "/usr/local/lib/python3.9/site-packages/requests/adapters.py", line 563, in send
pdns-admin-pgsql-1     |     raise SSLError(e, request=request)
pdns-admin-pgsql-1     | requests.exceptions.SSLError: HTTPSConnectionPool(host='authentik.lab.internal', port=443): Max retries exceeded with url: /application/o/powerdns-admin/.well-known/openid-configuration (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1129)')))
pdns-admin-pgsql-1     | [pid: 56|app: 0|req: 11804/11804] 127.0.0.1 () {60 vars in 1075 bytes} [Thu Sep 19 15:02:25 2024] GET /oidc/login => generated 6962 bytes in 85 msecs (HTTP/1.1 500) 3 headers in 215 bytes (1 switches on core 0)
```

**Comment by pschiffe** [comment]
Hi @thenetworkdoctor. What I think you have to do, is to add your CA to the system ca trust store, so the python will trust it. On a RHEL based system, this is done by copying your CA to the `/etc/pki/ca-trust/source/anchors/` dir (in the simple PEM or DER file formats) and running `update-ca-trust` command as root.

Could you try it? In a `pdns-admin` container. Given this is a Docker env, afterwards I would suggest to copy the content of the `/etc/pki/ca-trust/` dir to a docker volume, so the change will persist the container recreation.

**Comment by thenetworkdoctor** [comment]
@pschiffe , I tried something like this yesterday, I opened a shell to the container, installed ca-certificates, grep-ed the cert and imported it to the container, however this did not fix the issue. 

I'm guessing but i think the issue is the python flask webserver (who makes the call to my self hosted authentik) that doesn't use that cert dir.
Because of this i think i'll make an issue at the [Powerdns-admin](https://github.com/PowerDNS-Admin/PowerDNS-Admin) repo.
